geopandas/tools/_random.py

```diff
@@ -0,0 +1,83 @@
+from warnings import warn
+
+import numpy
+from shapely.geometry import MultiPoint
+
+from ..array import from_shapely, points_from_xy
+from ..geoseries import GeoSeries
+
+
+def uniform(geom, size, seed=None):
+    """
+
+    Sample uniformly at random from a geometry.
+
+    For polygons, this samples uniformly within the area of the polygon. For lines,
+    this samples uniformly along the length of the linestring. For multi-part
+    geometries, the weights of each part are selected according to their relevant
+    attribute (area for Polygons, length for LineStrings), and then points are
+    sampled from each part uniformly.
+
+    Any other geometry type (e.g. Point, GeometryCollection) are ignored, and an
+    empty MultiPoint geometry is returned.
+
+    Parameters
+    ----------
+    geom : any shapely.geometry.BaseGeometry type
+        the shape that describes the area in which to sample.
+
+    size : integer
+        an integer denoting how many points to sample
+
+    Returns
+    -------
+    shapely.MultiPoint geometry containing the sampled points
+
+    Examples
+    --------
+    >>> from shapely.geometry import box
+    >>> square = box(0,0,1,1)
+    >>> uniform(square, size=102) # doctest: +SKIP
+    """
+    generator = numpy.random.default_rng(seed=seed)
+
+    if geom is None or geom.is_empty:
+        return MultiPoint()
+
+    if geom.geom_type in ("Polygon", "MultiPolygon"):
+        return _uniform_polygon(geom, size=size, generator=generator)
+
+    if geom.geom_type in ("LineString", "MultiLineString"):
+        return _uniform_line(geom, size=size, generator=generator)
+
+    warn(
+        f"Sampling is not supported for {geom.geom_type} geometry type.",
+        UserWarning,
+        stacklevel=8,
+    )
+    return MultiPoint()
+
+
+def _uniform_line(geom, size, generator):
+    """
+    Sample points from an input shapely linestring
+    """
+
+    fracs = generator.uniform(size=size)
+    return from_shapely(geom.interpolate(fracs, normalized=True)).unary_union()
+
+
+def _uniform_polygon(geom, size, generator):
+    """
+    Sample uniformly from within a polygon using batched sampling.
+    """
+    xmin, ymin, xmax, ymax = geom.bounds
+    candidates = []
+    while len(candidates) < size:
+        batch = points_from_xy(
```

Comment:
I think this is fine as a first implementation, but this sampling is potentially quite wasteful, if you have a lot of points, and your first sample gets say 95% of size, you would only need to target for another 5% but this will then try to draw the full length of `size` again. (But this is perhaps better than drawing too few points by guessing how many sampled points will be accepted on the next iteration).

Reply:
It is tricky to get the heuristic right if we wanted to use different size. It super depends on the convexity of each polygon. Maybe a less wasteful option would be to go with a number larger that `size` initially to have a higher chance of hitting the `size` at one go. But I'd leave that for later if needed.

Reply:
Agree, sounds good

Reply:
Yes, I kept the chunk size constant as a first pass to be conservative.

Then think the statistically efficient method is to sample each round proportional to the hit rate times the remaining sample size, but that can cause the size of a round to get very large very quickly.